contrib/avro/src/python/pants/contrib/avro/tasks/avro_gen.py

```diff
@@ -0,0 +1,93 @@
+  """Compile Avro schema, protocol, and IDL files to Java code."""
+
+  @classmethod
+  def register_options(cls, register):
+    super(AvroJavaGenTask, cls).register_options(register)
+    # pylint: disable=no-member
+    cls.register_jvm_tool(register,
+                          'avro-tools',
+                          classpath=[JarDependency(org='org.apache.avro', name='avro-tools', rev=AVRO_REV)],)
+    register('--runtime-deps', advanced=True, type=list, fingerprint=True,
+             default=['//:avro-java-runtime'],
+             help='A list of specs pointing to dependencies of Avro generated code.')
+
+  def __init__(self, *args, **kwargs):
+    super(AvroJavaGenTask, self).__init__(*args, **kwargs)
+
+  def synthetic_target_type(self, target):
+    return JavaLibrary
+
+  def synthetic_target_extra_dependencies(self, target, target_workdir):
+    dependencies = OrderedSet()
+    for dep_spec in self.get_options().runtime_deps:
+      dep_address = Address.parse(dep_spec)
+      try:
+        self.context.build_graph.maybe_inject_address_closure(dep_address)
+        dependencies.add(self.context.build_graph.get_target(dep_address))
+      except AddressLookupError as e:
+        raise AddressLookupError('{}\n  referenced from {} scope'.format(e, self.options_scope))
+
+    dependencies.update(target.dependencies)
+    return dependencies
+
+  def synthetic_target_extra_exports(self, target, target_workdir):
+    return self.synthetic_target_extra_dependencies(target, target_workdir)
+
+  def is_gentarget(self, target):
+    return isinstance(target, JavaAvroLibrary)
+
+  def _avro(self, args):
+    result = self.runjava(classpath=self.tool_classpath('avro-tools'),
+                          main='org.apache.avro.tool.Main',
+                          args=args)
+    if result != 0:
+      raise TaskError('avro-tools failed with exit status {}'.format(result))
+
+  def _compile_schema(self, source, workdir):
+    self._avro(['compile', 'schema', source, workdir])
+
+  def _compile_protocol(self, source, workdir):
+    self._avro(['compile', 'protocol', source, workdir])
+
+  def _compile_idl(self, source, workdir):
+    idl_filename = os.path.basename(source)
+    idl_base_filename, _ = os.path.splitext(idl_filename)
+    protocol_file = os.path.join(workdir, idl_base_filename + '.avpr')
+    self._avro(['idl', source, protocol_file])
+    self._compile_protocol(protocol_file, workdir)
+
+  def execute_codegen(self, target, target_workdir):
+    for source in target.sources_relative_to_buildroot():
+      if source.endswith('.avsc'):
+        self._compile_schema(source, target_workdir)
+      elif source.endswith('.avpr'):
+        self._compile_protocol(source, target_workdir)
+      elif source.endswith('.avdl'):
+        self._compile_idl(source, target_workdir)
+      else:
+        raise TaskError('Pants only understands Avro schema (.avsc), protocol (.avpr), and IDL (.avdl) files.')
```

Comment:
Should indicate which file you matched in error here.

Reply:
Done.